core/src/docs/rfcs/3734_buffered_reader.md

```diff
@@ -0,0 +1,175 @@
+
+# Motivation
+
+The objective is to mitigate the IO overhead. In certain scenarios, the reader processes the data incrementally, meaning that it utilizes the `seek()` function to navigate to a specific position within the file. Subsequently, it invokes the `read()` to reads `limit` bytes into memory and performs the decoding process.
+
+
+OpenDAL triggers an IO request upon invoking `read()` if the `seek()` has reset the inner state. For storage services like S3, [research](https://www.vldb.org/pvldb/vol16/p2769-durner.pdf) suggests that an optimal IO size falls within the range of 8MiB to 16MiB. If the IO size is too small, the Time To First Byte (TTFB) dominates the overall runtime, resulting in inefficiency.
+
+Therefore, this RFC proposes the implementation of a buffered reader to amortize the overhead of IO.
+
+# Guide-level explanation
+
+For users who want to buffered reader, they will call the new API `buffer`. And the default behavior remains unchanged, so users using `op.reader_with()` are not affected. The `buffer` function will take a number as input, and this number will represent the maximum buffer capability the reader is able to use. 
+
+```rust
+op.reader_with(path).buffer(32 * 1024 * 1024).await
+```
+
+# Reference-level explanation
+
+This feature will be implemented in the `CompleteLayer`, with the addition of a `BufferReader` struct in `raw/oio/reader/buffer_reader.rs`. 
+
+The `BufferReader` employs` bytes::BytesMut` as the inner buffer and utilizes `(Option<usize>, Option<usize>)` to track the buffered range of the file.
+
+
+```rust
+     ...
+     async fn read(&self, path: &str, args: OpRead) -> Result<(RpRead, Self::Reader)> {
+          BufferReader::new(self.complete_read(path, args).await)
+     }
+
+     ...
+
+    fn blocking_read(&self, path: &str, args: OpRead) -> Result<(RpRead, Self::BlockingReader)> {
+          BufferReader::new(self.complete_blocking_read(path, args))
+    }
+     ...
+```
+
+A `buffer` field of type `Option<usize>` will be introduced to `OpRead`. If `buffer` is set to `None`, it functions with default behavior. However, if buffer is set to `Some(n)`, it denotes the maximum buffer capability that the `BufferReader` can utilize, and buffering behaviors introduced below.
+
+**Buffering**
+
+When the `Buffer Capability` is specified, the underlying reader fetches and buffers data in chunks corresponding to the buffer size.
+
+```
+Buffer
+┌───────┐
+│       │
+└───────┘
+File
+┌───┬───────┬──┐
+│   │       │  │
+└───▲───────▲──┘
+    │       │
+    │       │ 2. ReadToEnd(limit), The limit <= Cap(Buffer).
+    │
+    │ 1. SeekFromStart(128)
+
+```
+
+```
+Buffer
+┌───────┐
+│*******│
+└───────┘
+File
+┌───┬───────┬──┐
+│   │*******│  │
+└───▲───────▲──┘
+     3. Fetches and buffers the data.
+```
+
+**Tailing Read**
+
+If a read request attempts to read the trailing bytes of the file, it only reads and buffers the `Min(Max(Limit,Cap(Buffer)),FileEnd-Cursor)` bytes.
```

Comment:
Consider the following scenario:

The last four bytes represent the size of the metadata, and the code for reading the metadata might look like:

```rust
reader.seek(SeekFrom::End(-4));
let mut buf = [0u8; 4];
reader.read_exact(&mut buf)?;
let size = u32::from_le_bytes(buf);

reader.seek(SeekFrom::End(-(4 + size as i64)))?;
let mut buf = vec![0u8; size as usize];
reader.read_exact(&mut buf)?;
let metadata = parse_metadata(&buf)?;
```

According to the design presented here, would this lead to two separate calls to the underlying Seek and Read operations, rendering the use of a buffer ineffective?

Reply:
Yes, it will. I'm also considering introducing the tailing buffer feature #3745. Theoretically, this feature could reduce latency by around 25ms when reading trailing bytes from S3.

Reply:
Just to confirm, does the goal of this RFC include reducing the number of `seek` and `read` calls on the underlying storage?

Reply:
Yes, it is. Now the `read` of `RangeReader` is super expensive. 

> RangeReader will send request to read the entire range everytime no matter how large the read buffer size is. 

See also #3747

However, the `BufferReader` might not work well If the `read` always bypasses the buffer🥹.

Reply:
> Now the `read` of `RangeReader` is super expensive.

Correct: Both `read` and `seek` is not expensive, only the case that `read` after `seek` expensive. As explained in https://github.com/apache/incubator-opendal/issues/3747, we predict users to read the whole range, but they don't.

Reply:
How about implementing a basic buffer similar to `std::io::BufReader` and postponing the `tail read` feature? It seems that specific file formats are too complex for a reader (AsyncRead & AsyncSeek) to handle effectively.

Reply:
> How about implementing a basic buffer similar to `std::io::BufReader` and postponing the `tail read` feature? It seems that specific file formats are too complex for a reader (AsyncRead & AsyncSeek) to handle effectively.

Yes, the tailing buffer feature isn't part of this RFC.

Reply:
SO we don't need to discuss different cases one by one, right? How about refer to `std::io::BufReader` directly?

Reply:
> So we don't need to discuss different cases one by one, right? How about refer to `std::io::BufReader` directly?

OK, I will check the `std::io::BufReader` docs and implementation later to see if I need to add supplementary.

Reply:
@Xuanwo, I've checked the behaviors of `std::io::BufReader` and `tokio::io::BufferReader` to our RFC. The key distinction is that both `BufReader` implementations consistently discard the inner buffer when invoking the `seek` method.  My colleague @zhongzc also mentioned that its behavior should be more akin to the `Cache` than the `Buffer`. And it might be a little misleading if we also called our implementation `BufReader`.

Reply:
> And it might be a little misleading if we also called our implementation `BufReader`.

I'm OK with either names. They are not part of public API so users won't operate on buffer reader directly.

Reply:
So, back to the earlier suggestion, I think we can't simply refer to `std::io::BufReader`. It may be better to discuss different cases one by one. Would you happen to have any suggestions?

Reply:
> The key distinction is that both `BufReader` implementations consistently discard the inner buffer when invoking the `seek` method.

This understanding is incorrect: https://doc.rust-lang.org/std/io/struct.BufReader.html#method.seek_relative

Reply:
OK, I know there is `seek_relative`, but I think many readers invoke the `seek` and `read` methods, resulting in ineffective. https://doc.rust-lang.org/std/io/trait.Seek.html#tymethod.seek

Reply:
> OK, I know there is `seek_relative`, but I think many format readers just naively invoke the `seek` and `read` methods, resulting in inefficient.

Got it. OpenDAL's Reader doesn't provide `into_inner` like `BufReader`, so it's safe for us to always use seek_relative while ever possible.

Reply:
Yes. Let me convert the RFC to a draft and think about how to implement the feature we talked about above.